GitUI/CustomDiffMergeToolProvider.cs

```diff
@@ -0,0 +1,113 @@
+﻿using System.Collections.Generic;
+using System.ComponentModel;
+using System.Drawing;
+using System.Linq;
+using System.Windows.Forms;
+using GitCommands;
+
+namespace GitUI
+{
+    public class CustomDiffMergeToolProvider
+    {
+        /// <summary>
+        /// Time to wait before loading custom diff tools in FormBrowse
+        /// Try avoid loading while git-log and git-diff runs
+        /// </summary>
+        public readonly int FormBrowseToolDelay = 15000;
+
+        /// <summary>
+        /// Clear the existing caches
+        /// </summary>
+        public void Clear()
+        {
+            CustomDiffMergeToolCache.DiffToolCache.Clear();
+            CustomDiffMergeToolCache.MergeToolCache.Clear();
+        }
+
+        /// <summary>
+        /// Load the available  DiffMerge tools and apply to the menus
+        /// </summary>
+        /// <param name="module">The Git module</param>
+        /// <param name="menus">The menus to update</param>
+        /// <param name="components">The calling Form components, to dispose correctly</param>
+        /// <param name="isDiff">True if diff, false if merge</param>
+        /// <param name="delay">The delay before starting the operation</param>
+        public void LoadCustomDiffMergeTools(GitModule module, IList<CustomDiffMergeTool> menus, IContainer components, bool isDiff, int delay = 1000)
+        {
+            InitMenus(menus);
+
+            if (isDiff && !AppSettings.ShowAvailableDiffTools)
+            {
+                return;
+            }
+
+            ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            {
+                // get the tools, possibly with a delay as requesting requires considerable time
+                // cache is shared
+                IEnumerable<string> tools = await (isDiff ? CustomDiffMergeToolCache.DiffToolCache : CustomDiffMergeToolCache.MergeToolCache)
+                    .GetToolsAsync(module, delay);
```

Comment:
❗ `tools` enumerated multiple times, especially in the loop.
The collection must be materialised in a form or a list or an array

Reply:
CustomDiffMergeToolCache could return a List instead
Changed to ToList()